app/src/main/java/woowacourse/paint/BoardView.kt

```diff
@@ -54,28 +76,105 @@ class BoardView(context: Context, attrs: AttributeSet) : View(context, attrs) {
+        invalidate()
+    }
+
+    override fun onDraw(canvas: Canvas) {
+        super.onDraw(canvas)
+
+        drawingHistory.history.forEach { (path: Path, paint: Paint) ->
+            canvas.drawPath(path, paint)
+        }
+        setupPaint()
+        canvas.drawPath(path, paint)
+    }
+
     @SuppressLint("ClickableViewAccessibility")
     override fun onTouchEvent(event: MotionEvent): Boolean {
         when (event.action) {
-            MotionEvent.ACTION_DOWN -> startDrawing(event)
-            MotionEvent.ACTION_MOVE -> moveDrawing(event)
-            MotionEvent.ACTION_UP -> endDrawing()
+            MotionEvent.ACTION_DOWN -> startDrawing(event, brushType)
+            MotionEvent.ACTION_MOVE -> moveDrawing(event, brushType)
+            MotionEvent.ACTION_UP -> endDrawing(brushType)
         }
         invalidate()
         return true
     }
 
-    private fun startDrawing(event: MotionEvent) {
-        path.moveTo(event.x, event.y)
-        path.lineTo(event.x, event.y)
+    private fun startDrawing(event: MotionEvent, brushType: BrushType) {
+        when (brushType) {
+            BrushType.PEN -> {
+                path.moveTo(event.x, event.y)
+                path.lineTo(event.x, event.y)
+            }
+            BrushType.ERASER -> {
+                erasePath(event)
+            }
+            BrushType.ERASER_LINE -> {
+                path.moveTo(event.x, event.y)
+                path.lineTo(event.x, event.y)
+            }
+            else -> {
+                startX = event.x
+                startY = event.y
+            }
+        }
     }
 
-    private fun moveDrawing(event: MotionEvent) {
-        path.lineTo(event.x, event.y)
+    private fun erasePath(event: MotionEvent) {
+        val x = event.x
+        val y = event.y
+        val indexesToRemove = mutableListOf<Int>()
+
+        drawingHistory.history.forEachIndexed { index, drawing ->
+            val bounds = RectF()
+            drawing.path.computeBounds(bounds, true)
+            if (bounds.contains(x, y)) indexesToRemove.add(index)
+        }
+
+        indexesToRemove.asReversed().forEach {
+            drawingHistory.removeAt(it)
+        }
+
+        invalidate()
     }
 
-    private fun endDrawing() {
-        storedCanvas.drawPath(path, paint)
-        path.reset()
+    private fun moveDrawing(event: MotionEvent, brushType: BrushType) {
+        when (brushType) {
+            BrushType.PEN -> {
+                path.lineTo(event.x, event.y)
+            }
+            BrushType.RECTANGLE, BrushType.FILLED_RECTANGLE -> {
+                path.reset()
+                path.addRect(startX, startY, event.x, event.y, Path.Direction.CCW)
+                invalidate()
+            }
+            BrushType.CIRCLE, BrushType.FILLED_CIRCLE -> {
+                path.reset()
+                path.addCircle(startX, startY, event.x - startX, Path.Direction.CCW)
+                invalidate()
+            }
+            BrushType.ERASER -> {}
```

Comment:
빈 함수 처리는 아래와 같이 하면 직관적이고 보기 좋습니다!

```suggestion
BrushType.ERASER -> Unit
```